### PR TITLE
Implement STRIP Defense Against Poisoning Attacks

### LINKED ISSUE #664: Implement STRIP defense against poisoning attacks
**Is your feature request related to a problem? Please describe.**
Defense to abstain from classifying samples when there is poison. 

**Describe the solution you'd like**
Implementation of STRIP: A Defence Against Trojan Attacks on Deep Neural Networks 

**Describe alternatives you've considered**
NA

**Additional context**
NA

### PR DESCRIPTION
**Pull request by ebubae**
# Description

This PR:

- Implements the STRIP defense proposed by [Gao. et. al 2020](https://arxiv.org/abs/1902.06531).
- Creates a `poison_mitigation` folder to host mixins used for poison mitigation
- Resolves #664

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [x] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

Currently tests are running and pass for Keras and TF Keras. I had some issues with the Pytorch unit test classifier.

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: tests/defences/test_strip.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
+# MIT License
```
Comment (beat-buesser) [inline tests/defences/test_strip.py:1]: Could you please convert these tests to the new pytest pattern: https://github.com/Trusted-AI/adversarial-robustness-toolbox/wiki/ART-Unit-Testing

File: run_tests.sh [hunk 2]
```diff
@@ -111,6 +111,7 @@ declare -a defences=("tests/defences/test_adversarial_trainer.py" \
                      "tests/defences/test_pixel_defend.py" \
                      "tests/defences/test_reverse_sigmoid.py" \
                      "tests/defences/test_rounded.py" \
+                     "tests/defences/test_strip.py" \
```
Comment (beat-buesser) [inline run_tests.sh:114]: Could you please convert these tests to the new pytest pattern: https://github.com/Trusted-AI/adversarial-robustness-toolbox/wiki/ART-Unit-Testing

File: art/defences/transformer/poisoning/strip.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        return self.classifier
+
+    def fit(self, x: np.ndarray, y: Optional[np.ndarray] = None, **kwargs) -> None:
+        """
+        No parameters to learn for this method; do nothing.
+        """
+        raise NotImplementedError
+
+    def _check_params(self) -> None:
+        pass
```
Comment (beat-buesser) [inline art/defences/transformer/poisoning/strip.py:90]: Could you please add checks for each of the arguments of `__init__`.

File: art/defences/transformer/poisoning/strip.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        :param classifier: A trained classifier.
+        """
+        super().__init__(classifier=classifier)
+        self._check_params()
+
+    def __call__(
+            self,
+            num_samples: int = 20,
+            false_acceptance_rate: float = 0.01,
+    ) -> "CLASSIFIER_TYPE":
```
Comment (beat-buesser) [inline art/defences/transformer/poisoning/strip.py:64]: Could we make the return type more specific? I think it's not just a general classifier anymore.

File: art/estimators/poison_mitigation/strip/strip.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+| Paper link: https://arxiv.org/abs/1902.06531
+"""
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import logging
+import numpy as np
+from typing import Callable, Optional
+
+from scipy.stats import entropy, norm
+from tqdm import tqdm
```
Comment (beat-buesser) [inline art/estimators/poison_mitigation/strip/strip.py:30]:
```suggestion
import logging
from typing import Callable, Optional

import numpy as np
from scipy.stats import entropy, norm
from tqdm import tqdm
```

File: art/estimators/poison_mitigation/strip/strip.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+        num_samples: int = 20,
+        false_acceptance_rate: float = 0.01,
+        **kwargs
+    ) -> None:
+        """
+        Create a STRIP defense
+
+        :param num_samples: The number of samples to use to test entropy at inference time
+        :param false_acceptance_rate: The percentage of acceptable false acceptance
+        :param predict function
```
Comment (beat-buesser) [inline art/estimators/poison_mitigation/strip/strip.py:55]: I think this line needs an update.

File: art/estimators/poison_mitigation/strip/strip.py [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+            return raw_predictions
+
+        x_val = self.validation_data
+        final_predictions = []
+
+        for i, img in enumerate(x):
+            # Randomly select samples from test set
+            selected_indices = np.random.choice(np.arange(len(x_val)), self.num_samples)
+
+            # Perturn the images by combining them
```
Comment (beat-buesser) [inline art/estimators/poison_mitigation/strip/strip.py:84]:
```suggestion
            # Perturb the images by combining them
```

**Comment by beat-buesser** [comment]
Could you please update `art.estimators.__init__.py` with an import of `poison_mitigation` after line 18?

File: art/estimators/poison_mitigation/strip/__init__.py [hunk 8]
```diff
@@ -0,0 +1,4 @@
+"""
+Neural cleanse estimators.
```
Comment (beat-buesser) [inline art/estimators/poison_mitigation/strip/__init__.py:2]: I think this docstring needs an update.

**Review by beat-buesser — CHANGES_REQUESTED** [review]
Hi @ebubae Thank you very much for implementing a new defence against poisoning! A think it's a very nice implementation and I only have a few formatting requests.

File: tests/defences/test_strip.py (hunk 1 above)
Comment (ebubae) [reply tests/defences/test_strip.py:1]: Done

File: run_tests.sh (hunk 2 above)
Comment (ebubae) [reply run_tests.sh:103]: Done. Please check that it's truly framework-independent. I'm not sure if it's only testing TF.

File: art/defences/transformer/poisoning/strip.py (hunk 4 above)
Comment (ebubae) [reply art/defences/transformer/poisoning/strip.py]: This is a very good point. I've added a `TypeVar` to bound the types.

File: art/estimators/poison_mitigation/strip/strip.py (hunk 6 above)
Comment (ebubae) [reply art/estimators/poison_mitigation/strip/strip.py]: Done

File: art/estimators/poison_mitigation/strip/strip.py (hunk 7 above)
Comment (ebubae) [reply art/estimators/poison_mitigation/strip/strip.py]: Thanks for catching this.

File: art/estimators/poison_mitigation/strip/__init__.py (hunk 8 above)
Comment (ebubae) [reply art/estimators/poison_mitigation/strip/__init__.py]: Fixed

**Review by beat-buesser — APPROVED** [review]
Hi @ebubae Thank you very much for implementing the STRIP defense against poisoning attacks!

**Pull request merged**